### PR TITLE
Fix: Treat Generic classes as not being is_list_like

### LINKED ISSUE #49649: BUG: warning from typing library if you subclass a DataFrame using Generic with python 3.11
### Pandas version checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pandas.pydata.org/docs/whatsnew/index.html) of pandas.

- [ ] I have confirmed this bug exists on the main branch of pandas.

### Reproducible Example

```python
import pandas as pd
from typing import TypeVar, Generic

T = TypeVar("T")

class MyDataFrame(pd.DataFrame, Generic[T]):
    ...

def func() -> MyDataFrame[int]:
    return MyDataFrame[int]({"foo": [1, 2, 3]})

x = func()
print(x)
```

### Issue Description

With pandas 1.5.1 and python 3.10, the code above gives no warning at runtime.

With pandas 1.5.1 and python 3.11, you get a warning from the typing library at runtime:
```text
C:\Anaconda3\envs\pandasstubs311\Lib\typing.py:1253: UserWarning: Pandas doesn't allow columns to be created via a new attribute name - see https://pandas.pydata.org/pandas-docs/stable/indexing.html#attribute-access
  result.__orig_class__ = self
```

We have testing code in pandas-stubs that uses this, because it was reported by the people developing the `pandera` library.

### Expected Behavior

No warning.

We're doing something in pandas that doesn't allow a `Generic` subclass to be created with python 3.11.

### Installed Versions

<details>

INSTALLED VERSIONS
------------------
commit           : 91111fd99898d9dcaa6bf6bedb662db4108da6e6
python           : 3.11.0.final.0
python-bits      : 64
OS               : Windows
OS-release       : 10
Version          : 10.0.19043
machine          : AMD64
processor        : Intel64 Family 6 Model 158 Stepping 13, GenuineIntel
byteorder        : little
LC_ALL           : None
LANG             : None
LOCALE           : English_United States.1252

pandas           : 1.5.1
numpy            : 1.23.4
pytz             : 2022.6
dateutil         : 2.8.2
setuptools       : 65.5.1
pip              : 22.3.1
Cython           : None
pytest           : 7.2.0
hypothesis       : None
sphinx           : None
blosc            : None
feather          : None
xlsxwriter       : None
lxml.etree       : None
html5lib         : 1.1
pymysql          : None
psycopg2         : None
jinja2           : 3.1.2
IPython          : None
pandas_datareader: None
bs4              : None
bottleneck       : None
brotli           : None
fastparquet      : None
fsspec           : None
gcsfs            : None
matplotlib       : 3.6.2
numba            : None
numexpr          : None
odfpy            : None
openpyxl         : 3.0.10
pandas_gbq       : None
pyarrow          : None
pyreadstat       : None
pyxlsb           : 1.0.10
s3fs             : None
scipy            : 1.9.3
snappy           : None
sqlalchemy       : 1.4.43
tables           : None
tabulate         : 0.9.0
xarray           : 2022.11.0
xlrd             : 2.0.1
xlwt             : None
zstandard        : None
tzdata           : None
</details>

### PR DESCRIPTION
**Pull request by Dr-Irv**
- [x] closes #49649
- [x] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
  - `tests/dtypes/test_inference.py:test_is_list_like_generic`
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [x] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [x] Added an entry in the latest `doc/source/whatsnew/v1.5.2.rst` file if fixing a bug or adding a new feature.

### CODE REVIEW

**Comment by WillAyd** [comment]
Isn't python 3.11 exhibiting the correct behavior?

**Comment by Dr-Irv** [comment]
> Isn't python 3.11 exhibiting the correct behavior?

No, see the example in #49649 .  It is preventing subclassing `DataFrame` as a `Generic`.

I think what happened here is that python 3.11 added an `__iter__()` method to the class representing the type of the `Generic`, so we see that type as being "list like".  Then internally, the typing module assigns an attribute to the instance, which we then complain about.  Prior to 3.11, the `__iter__()` method wasn't there, so we were OK.

**Comment by WillAyd** [comment]
Ah OK thanks for the explanation. If you replaced `Gen[T]` in your test with `list[T]` what would your expectation be? I think we should add that as a test as well

**Comment by Dr-Irv** [comment]
> Ah OK thanks for the explanation. If you replaced `Gen[T]` in your test with `list[T]` what would your expectation be? I think we should add that as a test as well

Before seeing that message, I updated the test to more explicitly test the case of subclassing a `DataFrame` .  So I'm not sure if we need to test `list[T]` .  

The issue is that when you instantiate the subclass that is `Generic`, the typing module creates an attribute on the class and we issue a warning:
```text
C:\Anaconda3\envs\pandas-dev311\Lib\typing.py:1253: UserWarning: Pandas doesn't allow columns to be created via a new attribute name - see https://pandas.pydata.org/pandas-docs/stable/indexing.html#attribute-access
  result.__orig_class__ = self
```

Here `result` is a `DataFrame`, and `self` is the `_GenericAlias` .

File: pandas/tests/dtypes/test_inference.py [hunk 1]
```diff
@@ -229,6 +233,21 @@ def __getitem__(self, item):
...
+    T = TypeVar("T")
+
+    class MyDataFrame(DataFrame, Generic[T]):
+        ...
+
+    tstc = MyDataFrame[int]
+    tst = MyDataFrame[int]({"x": [1, 2, 3]})
+
+    assert not inference.is_list_like(tstc)
+    assert isinstance(tst, DataFrame)
```
Comment (twoertwein) [inline pandas/tests/dtypes/test_inference.py:248]: Might be worth also adding this test (DataFrames are list-like): `assert inference.is_list_like(tst)`

File: pandas/_libs/lib.pyx [hunk 2]
```diff
@@ -1100,6 +1103,8 @@ cdef inline bint c_is_list_like(object obj, bint allow_sets) except -1:
         and not (hasattr(obj, "ndim") and obj.ndim == 0)
         # exclude sets if allow_sets is False
         and not (allow_sets is False and isinstance(obj, abc.Set))
+        # exclude Generic types that have __iter__
+        and not isinstance(obj, _GenericAlias)
```
Comment (twoertwein) [inline pandas/_libs/lib.pyx:1107]: Coul also combine it with the earlier exception for `str` and `bytes`

Comment (twoertwein) [reply pandas/_libs/lib.pyx:1107]: Is there a public class to test against?

Comment (Dr-Irv) [reply pandas/_libs/lib.pyx]: > Coul also combine it with the earlier exception for `str` and `bytes`

done in https://github.com/pandas-dev/pandas/pull/49736/commits/5260e4cff5076c6aaeb4883a0835988e74d2a7e3

File: pandas/tests/dtypes/test_inference.py (hunk 1 above)
Comment (Dr-Irv) [reply pandas/tests/dtypes/test_inference.py:248]: > Might be worth also adding this test (DataFrames are list-like): `assert inference.is_list_like(tst)`

done in https://github.com/pandas-dev/pandas/pull/49736/commits/a36d1c066f0f31196d0f8318c5a98a69246e593e

File: pandas/_libs/lib.pyx (hunk 2 above)
Comment (twoertwein) [reply pandas/_libs/lib.pyx]: Can we simply test against `type`: `not isinstance(obj, type)`? Not sure whether that might be too wide.

Comment (Dr-Irv) [reply pandas/_libs/lib.pyx]: > Can we simply test against `type`: `not isinstance(obj, type)`? Not sure whether that might be too wide.

No.  See below.
```python
>>> from typing import TypeVar, Generic
>>> T = TypeVar("T")
>>> class MyDataFrame(DataFrame, Generic[T]):
...     ...
...
>>> tstc = MyDataFrame[int]
>>> isinstance(tstc, type)
False
>>> from typing import _GenericAlias
>>> isinstance(tstc, _GenericAlias)
True
```

**Review by WillAyd — APPROVED** [review]
lgtm

File: doc/source/whatsnew/v1.5.2.rst [hunk 3]
```diff
@@ -27,7 +27,7 @@ Fixed regressions
 Bug fixes
 ~~~~~~~~~
 - Bug in the Copy-on-Write implementation losing track of views in certain chained indexing cases (:issue:`48996`)
--
+- Bug with python 3.11 and creating a subclass that is ``Generic`` (:issue:`49649`)
```
Comment (twoertwein) [inline doc/source/whatsnew/v1.5.2.rst:30]: Maybe make it clearer that it is `typing.Generic`:

`Fixed bug when instantiating a ``DataFrame`` subclass inheriting from ``typing.Generic`` that triggered a ``UserWarning`` on python 3.11 (:issue:`49649`)`

Comment (Dr-Irv) [reply doc/source/whatsnew/v1.5.2.rst]: Good suggestion.  Fixed in https://github.com/pandas-dev/pandas/pull/49736/commits/c62461db793d5cd7038b5b0cb43a931b6eb45038

**Comment by jbrockmendel** [comment]
Perf?

**Comment by Dr-Irv** [comment]
> Perf?

Have to assume that the difference between doing `isinstance(obj, (str, bytes))` and `isinstance(obj, (str, bytes, _GenericAlias))` is minimal.  And that our benchmarks won't see the difference.

**Comment by jbrockmendel** [comment]
I'm sure it will be small, but this is a function that we've gone out of our way to optimize to the gills

**Comment by Dr-Irv** [comment]
> I'm sure it will be small, but this is a function that we've gone out of our way to optimize to the gills

So what's the methodology to do the comparison?  In CI?  Locally?

And if there is a hit, then what do we do about it?

**Comment by jbrockmendel** [comment]
> So what's the methodology to do the comparison? In CI? Locally?

usually show some before/after results with `%timeit` in a few cases chosen using your best judgement

**Comment by Dr-Irv** [comment]
> > So what's the methodology to do the comparison? In CI? Locally?
> 
> usually show some before/after results with `%timeit` in a few cases chosen using your best judgement

OK, so I did this test
BEFORE
```python
%timeit inference.is_list_like(set([1,2,3]))
572 ns ± 14.4 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)
```

AFTER
```python
%timeit inference.is_list_like(set([1,2,3]))
592 ns ± 12 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)
```

I tested a `set` because it is the last thing checked in `is_list_like`, so we know the new test is always executed.  This is with python 3.8 on my Windows machine.

Open to other tests to try.

**Comment by jbrockmendel** [comment]
thanks

File: doc/source/whatsnew/v1.5.2.rst [hunk 4]
```diff
@@ -27,8 +27,8 @@ Fixed regressions
 Bug fixes
 ~~~~~~~~~
 - Bug in the Copy-on-Write implementation losing track of views in certain chained indexing cases (:issue:`48996`)
+- Fixed bug when instantiating a ``DataFrame`` subclass inheriting from ``typing.Generic`` that triggered a ``UserWarning`` on python 3.11 (:issue:49649)
```
Comment (mroeschke) [inline doc/source/whatsnew/v1.5.2.rst:30]:
```suggestion
- Fixed bug when instantiating a :class:`DataFrame` subclass inheriting from ``typing.Generic`` that triggered a ``UserWarning`` on python 3.11 (:issue:`49649`)
```

**Comment by Dr-Irv** [comment]
> Looks like the whatsnew note will need to move to 1.5.3

That's unfortunate.  I was hoping this would get in to clean up some test code in `pandas-stubs`.  Oh, well

**Comment by Dr-Irv** [comment]
> Looks like the whatsnew note will need to move to 1.5.3

Actually, who's responsible for creating the new 1.5.3 whatsnew template?

**Comment by mroeschke** [comment]
I _think_ the new whatsnew template is done at the end of the release process? cc @datapythonista

**Comment by datapythonista** [comment]
> > Looks like the whatsnew note will need to move to 1.5.3
> 
> Actually, who's responsible for creating the new 1.5.3 whatsnew template?

Yes, I'll be doing it shortly, sorry for the delay.

**Comment by WillAyd** [comment]
Maybe worth updating the branch again but otherwise lgtm to merge

**Comment by Dr-Irv** [comment]
> Maybe worth updating the branch again but otherwise lgtm to merge

I moved the whatsnew comment from 1.5.2 to 1.5.3 in https://github.com/pandas-dev/pandas/pull/49736/commits/a4142ad6175c279af2a0d4321d0fcfedba18e7c1

**Comment by mroeschke** [comment]
Sorry looks like another merge conflict slipped in.

Also the diff shows that `doc/source/whatsnew/v1.5.2.rst` is still modified (errant empty note)

**Comment by Dr-Irv** [comment]
> Sorry looks like another merge conflict slipped in.
> 
> Also the diff shows that `doc/source/whatsnew/v1.5.2.rst` is still modified (errant empty note)

Should be fixed in https://github.com/pandas-dev/pandas/pull/49736/commits/cb6140f5923c1c1d72ef77dfba0bd1674fc961e8

**Review by mroeschke — APPROVED** [review]
Thanks for the quick fix. I'll merge on green

**Comment by mroeschke** [comment]
Thanks @Dr-Irv

**Pull request merged**